lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,442 @@
+            GitError::Subprocess(error) => GitPushError::Subprocess(error),
+        }
+    }
+}
+
+/// Context for creating git subprocesses
+pub(crate) struct GitSubprocessContext<'a> {
+    git_dir: PathBuf,
+    work_tree: PathBuf,
+    git_path: &'a Path,
+}
+
+fn canonicalize(path: impl AsRef<Path>) -> Result<PathBuf, GitSubprocessError> {
+    dunce::canonicalize(&path)
+        .map_err(|e| GitSubprocessError::PathCanonicalization(path.as_ref().to_path_buf(), e))
+}
+
+impl<'a> GitSubprocessContext<'a> {
+    pub(crate) fn new(
+        git_dir: impl Into<PathBuf>,
+        work_tree: impl Into<PathBuf>,
+        git_path: &'a Path,
+    ) -> Self {
+        GitSubprocessContext {
+            git_dir: git_dir.into(),
+            work_tree: work_tree.into(),
+            git_path,
+        }
+    }
+
+    pub(crate) fn from_git2(
+        git_repo: &git2::Repository,
+        git_path: &'a Path,
+    ) -> Result<Self, GitError> {
+        let git_dir = canonicalize(git_repo.path())?;
+        let work_tree = work_tree_from_git_dir(git_repo.path())
+            .map_err(GitSubprocessError::WorkTreePath)
+            .map(canonicalize)??;
+
+        Ok(Self::new(git_dir, work_tree, git_path))
+    }
+
+    fn spawn<I, S>(&self, args: I, stdout: Stdio) -> Result<Output, GitError>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let mut git_cmd = Command::new(self.git_path);
+        // we cd into the work tree to avoid
+        git_cmd
+            .args([
+                OsStr::new("--git-dir"),
+                self.git_dir.as_ref(),
+                OsStr::new("--work-tree"),
+                self.work_tree.as_ref(),
+            ])
+            .args(args)
+            .stdin(Stdio::null())
+            .stdout(stdout)
+            .stderr(Stdio::piped());
+        tracing::debug!(cmd = ?git_cmd, "spawning a git subprocess");
+        let child_git = git_cmd.spawn().map_err(|e| {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                GitError::GitCommandNotFound(
+                    (self.git_path != Path::new("git")).then_some(self.git_path.to_path_buf()),
+                )
+            } else {
+                GitSubprocessError::Fork(e).into()
+            }
+        })?;
+
+        let output = child_git
+            .wait_with_output()
+            .map_err(GitSubprocessError::Wait)?;
+        Ok(output)
+    }
+
+    /// Perform a git fetch
+    pub(crate) fn spawn_fetch(
+        &self,
+        remote_name: &str,
+        depth: Option<NonZeroU32>,
+        refspec: &str,
+        prunes: &mut Vec<String>,
+    ) -> Result<(), GitError> {
+        let depth_arg = depth.map(|x| format!("--depth={x}"));
+        let args = {
+            let mut a = vec!["fetch".to_string(), "--prune".to_string()];
+            if let Some(depth) = depth_arg {
+                a.push(depth);
+            }
```

Comment:
nit: since we're checking if `depth_arg` is `Some()` here, it's simpler to not do the `map()` call above and instead make this look like this:

```rust
if let Some(depth) = depth {
    a.push(format!("--depth={depth}");
}
```